aiohttp/web.py

```diff
@@ -1195,35 +1355,45 @@ def handle_request(self, message, payload):
 
         app = self._app
         request = Request(app, message, payload,
-                          self.transport, self.writer, self.keep_alive_timeout)
+                          self.transport, self.reader, self.writer,
+                          self.keep_alive_timeout)
         try:
-            match_info = yield from self._router.resolve(request)
+            try:
+                match_info = yield from self._router.resolve(request)
 
-            assert isinstance(match_info, AbstractMatchInfo), match_info
+                assert isinstance(match_info, AbstractMatchInfo), match_info
 
-            request._match_info = match_info
-            handler = match_info.handler
+                request._match_info = match_info
+                handler = match_info.handler
 
-            for factory in reversed(self._middlewares):
-                handler = yield from factory(app, handler)
-            resp = yield from handler(request)
+                for factory in reversed(self._middlewares):
+                    handler = yield from factory(app, handler)
+                resp = yield from handler(request)
 
-            if not isinstance(resp, StreamResponse):
-                raise RuntimeError(
-                    ("Handler {!r} should return response instance, got {!r} "
-                     "[middlewares {!r}]")
-                    .format(match_info.handler, type(resp), self._middlewares))
-        except HTTPException as exc:
-            resp = exc
+                if not isinstance(resp, StreamResponse):
+                    raise RuntimeError(
+                        ("Handler {!r} should return response instance, "
+                         "got {!r} [middlewares {!r}]").format(
+                             match_info.handler,
+                             type(resp),
+                             self._middlewares))
+            except HTTPException as exc:
+                resp = exc
 
-        resp_msg = resp.start(request)
-        yield from resp.write_eof()
+            resp_msg = resp.start(request)
+            yield from resp.write_eof()
+
+            # notify server about keep-alive
+            self.keep_alive(resp_msg.keep_alive())
 
-        # notify server about keep-alive
-        self.keep_alive(resp_msg.keep_alive())
+            # log access
+            self.log_access(message, None, resp_msg, self._loop.time() - now)
 
-        # log access
-        self.log_access(message, None, resp_msg, self._loop.time() - now)
+        except DisconnectedError:
```

Comment:
1. this will catch disconnects from aiohttp.client
2. `ServerHttpProtocol.start()` already handles ClientDisconnectedError, change start()